File: go/ccxt/exchange_generic.go
package ccxt

import (
	"fmt"
	"reflect"
	"sort"
	"strconv"
)

func (this *Exchange) SortBy(array interface{}, value1 interface{}, desc2 ...interface{}) []interface{} {
	var desc bool
	var defaultValue interface{} = "a"
	if len(desc2) > 0 {
		desc = desc2[0].(bool)
	}
	list := array.([]interface{})

	if str, ok := value1.(string); ok {
		sort.Slice(list, func(i, j int) bool {
			a := list[i].(map[string]interface{})[str]
			b := list[j].(map[string]interface{})[str]
			return fmt.Sprintf("%v", a) < fmt.Sprintf("%v", b)
		})
		if desc {
			for i := len(list)/2 - 1; i >= 0; i-- {
				opp := len(list) - 1 - i
				list[i], list[opp] = list[opp], list[i]
			}
		}
		return list
	} else {
		value := value1.(int)
		sort.Slice(list, func(i, j int) bool {
			var a, b interface{}
			if reflect.TypeOf(list[i]).Kind() == reflect.Slice {
				a = list[i].([]interface{})[value]
			} else {
				a = defaultValue
			}
			if reflect.TypeOf(list[j]).Kind() == reflect.Slice {
				b = list[j].([]interface{})[value]
			} else {
				b = defaultValue
			}
			// return fmt.Sprintf("%v", a) < fmt.Sprintf("%v", b)
			switch aVal := a.(type) {
			case int:
				if bVal, ok := b.(int); ok {
					return aVal < bVal
				}
			case float64:
				if bVal, ok := b.(float64); ok {
					return aVal < bVal
				}
			case string:
				if bVal, ok := b.(string); ok {
					return aVal < bVal
				}
			}

			// Fallback to string comparison
			return fmt.Sprintf("%v", a) < fmt.Sprintf("%v", b)
		})
		if desc {
			// for i := len(list)/2 - 1; i >= 0; i-- {
			// 	opp := len(list) - 1 - i
			// 	list[i], list[opp] = list[opp], list[i]
			// }
			for i := 0; i < len(list)/2; i++ {
				opp := len(list) - 1 - i
				list[i], list[opp] = list[opp], list[i]
			}
		}
		return list
	}
}

func (this *Exchange) SortBy2(array interface{}, key1 interface{}, key2 interface{}, desc2 ...interface{}) []interface{} {
	var desc bool
	if len(desc2) > 0 {
		desc = desc2[0].(bool)
	}
	list := array.([]interface{})

	if str, ok := key1.(string); ok {
		key2Str, _ := key2.(string)
		sort.Slice(list, func(i, j int) bool {
			a1 := list[i].(map[string]interface{})[str]
			a2 := list[i].(map[string]interface{})[key2Str]
			b1 := list[j].(map[string]interface{})[str]
			b2 := list[j].(map[string]interface{})[key2Str]
			if a1 == b1 {
				return fmt.Sprintf("%v", a2) < fmt.Sprintf("%v", b2)
			}
			return fmt.Sprintf("%v", a1) < fmt.Sprintf("%v", b1)
		})
		if desc {
			for i := len(list)/2 - 1; i >= 0; i-- {
				opp := len(list) - 1 - i
				list[i], list[opp] = list[opp], list[i]
			}
		}
		return list
	}
	return nil
}

func (this *Exchange) FilterBy(aa interface{}, key interface{}, value interface{}) []interface{} {
	var targetA []interface{}
	if aaArr, ok := aa.([]interface{}); ok {
		targetA = aaArr
	} else {
		for _, v := range aa.(map[string]interface{}) {
			targetA = append(targetA, v)
		}
	}
	var outList []interface{}
	for _, elem := range targetA {
		if elem.(map[string]interface{})[key.(string)] == value {
			outList = append(outList, elem)
		}
	}
	return outList
}

func (this *Exchange) Extend(aa interface{}, bb ...interface{}) map[string]interface{} {
	return ExtendMap(aa, bb...)
}

func ExtendMap(aa interface{}, bb ...interface{}) map[string]interface{} {
	a := aa.(map[string]interface{})
	outDict := make(map[string]interface{})
	for key, value := range a {
		outDict[key] = value
	}
	if len(bb) > 0 {
		b, ok := bb[0].(map[string]interface{})
		if ok {
			for key, value := range b {
				outDict[key] = value
			}
		}
	}
	return outDict
}

func (this *Exchange) DeepExtend2(objs ...interface{}) interface{} {
	outDict := make(map[string]interface{})
	for _, obj := range objs {
		if obj == nil {
			obj = make(map[string]interface{})
		}
		if reflect.TypeOf(obj).Kind() == reflect.Map {
			for key, value := range obj.(map[string]interface{}) {
				if value != nil && reflect.TypeOf(value).Kind() == reflect.Map {
					if _, exists := outDict[key]; exists {
						outDict[key] = this.DeepExtend2(outDict[key], value)
					} else {
						outDict[key] = this.DeepExtend2(value)
					}
				} else {
					outDict[key] = value
				}
			}
		} else {
			outDict = obj.(map[string]interface{})
		}
	}
	return outDict
}

func (this *Exchange) DeepExtend(objs ...interface{}) map[string]interface{} {
	var outObj interface{}
	for _, x := range objs {
		if x == nil {
			continue
		}
		if reflect.TypeOf(x).Kind() == reflect.Map {
			if outObj == nil || reflect.TypeOf(outObj).Kind() != reflect.Map {
				outObj = make(map[string]interface{})
			}
			dictX := x.(map[string]interface{})
			for k, _ := range dictX {
				arg1 := outObj.(map[string]interface{})[k]
				arg2 := dictX[k]
				if arg1 != nil && arg2 != nil && reflect.TypeOf(arg1).Kind() == reflect.Map && reflect.TypeOf(arg2).Kind() == reflect.Map {
					outObj.(map[string]interface{})[k] = this.DeepExtend(arg1, arg2)
				} else {
					if arg2 != nil {
						outObj.(map[string]interface{})[k] = arg2
					} else {
						outObj.(map[string]interface{})[k] = arg1
					}
				}
			}
		} else {
			outObj = x
		}
	}
	return outObj.(map[string]interface{})
}

// func (this *Exchange) DeepExtend(objs ...interface{}) map[string]interface{} {
// 	var outObj map[string]interface{}

// 	for _, x := range objs {
// 		if x == nil {
// 			continue
// 		}

// 		dictX, ok := x.(map[string]interface{})
// 		if !ok {
// 			continue
// 		}

// 		if outObj == nil {
// 			outObj = make(map[string]interface{})
// 		}

// 		for k, v := range dictX {
// 			if existingVal, exists := outObj[k]; exists {
// 				if existingMap, ok1 := existingVal.(map[string]interface{}); ok1 {
// 					if vMap, ok2 := v.(map[string]interface{}); ok2 {
// 						// Recursively merge maps
// 						outObj[k] = this.DeepExtend(existingMap, vMap)
// 						continue
// 					}
// 				}
// 			}
// 			// Directly assign the value if no deep merging is needed
// 			outObj[k] = v
// 		}
// 	}

// 	return outObj
// }

// func (this *Exchange) DeepExtend(objs ...interface{}) map[string]interface{} {
// 	var outObj interface{}
// 	for _, x := range objs {
// 		if x == nil {
// 			continue
// 		}
// 		// if xMap, ok := x.(map[string]interface{}); ok {
// 		if xMap, ok := x.(map[string]interface{}); ok {
// 			if outObj == nil {
// 				outObj = make(map[string]interface{})
// 			} else if _, ok := x.(map[string]interface{}); !ok {
// 				//  || reflect.TypeOf(outObj).Kind() != reflect.Map
// 				outObj = make(map[string]interface{})
// 			}
// 			dictX := xMap
// 			for k, _ := range dictX {
// 				arg1 := outObj.(map[string]interface{})[k]
// 				arg2 := dictX[k]
// 				if arg1 != nil && arg2 != nil {
// 					_, arg1IsMap := arg1.(map[string]interface{})
// 					_, arg2IsMap := arg2.(map[string]interface{})
// 					if arg1IsMap && arg2IsMap {
// 						outObj.(map[string]interface{})[k] = this.DeepExtend(arg1, arg2)
// 					}
// 				} else {
// 					if arg2 != nil {
// 						outObj.(map[string]interface{})[k] = arg2
// 					} else {
// 						outObj.(map[string]interface{})[k] = arg1
// 					}
// 				}
// 			}
// 		} else {
// 			outObj = x
// 		}
// 	}
// 	return outObj.(map[string]interface{})
// }

// func (this *Exchange) DeepExtend(objs ...interface{}) map[string]interface{} {
// 	var outObj interface{}
// 	for _, x := range objs {
// 		if x == nil {
// 			continue
// 		}
// 		// if xMap, ok := x.(map[string]interface{}); ok {
// 		if xMap, ok := x.(map[string]interface{}); ok {
// 			if outObj == nil {
// 				outObj = make(map[string]interface{})
// 			} else if _, ok := x.(map[string]interface{}); !ok  {
// 				//  || reflect.TypeOf(outObj).Kind() != reflect.Map
// 				outObj = make(map[string]interface{})
// 			}
// 			dictX := xMap
// 			for k, _ := range dictX {
// 				arg1 := outObj.(map[string]interface{})[k]
// 				arg2 := dictX[k]
// 				if arg1 != nil && arg2 != nil {
// 					_, arg1IsMap := arg1.(map[string]interface{})
// 					_, arg2IsMap := arg2.(map[string]interface{})
// 					if arg1IsMap && arg2IsMap {
// 						outObj.(map[string]interface{})[k] = this.DeepExtend(arg1, arg2)
// 					}
// 				} else {
// 					if arg2 != nil {
// 						outObj.(map[string]interface{})[k] = arg2
// 					} else {
// 						outObj.(map[string]interface{})[k] = arg1
// 					}
// 				}
// 			}
// 		} else {
// 			outObj = x
// 		}
// 	}
// 	return outObj.(map[string]interface{})
// }

// func (this *Exchange) InArray(elem interface{}, list2 interface{}) bool {
// 	if list2 == nil {
// 		return false
// 	}
// 	if reflect.TypeOf(list2).Kind() == reflect.Slice {
// 		list := list2.([]interface{})
// 		for _, v := range list {
// 			if v == elem {
// 				return true
// 			}
// 		}
// 	}
// 	return false
// }

func (this *Exchange) InArray(elem interface{}, list interface{}) bool {
	// Ensure the list is not nil and is of a slice type
	if list == nil || reflect.TypeOf(list).Kind() != reflect.Slice {
		return false
	}

	// Use reflection to iterate over the slice
	listValue := reflect.ValueOf(list)
	for i := 0; i < listValue.Len(); i++ {
		listElem := listValue.Index(i).Interface()

		// Handle number comparison
		switch e := elem.(type) {
		case int, int8, int16, int32, int64, float32, float64:
			switch l := listElem.(type) {
			case int, int8, int16, int32, int64, float32, float64:
				// Convert both to float64 for comparison
				if ToFloat64(e) == ToFloat64(l) {
					return true
				}
			}
		default:
			// For non-numeric values, use DeepEqual
			if reflect.DeepEqual(listElem, elem) {
				return true
			}
		}
	}

	return false
}

// func (this *Exchange) InArray(elem interface{}, list interface{}) bool {
// 	// Ensure the list is not nil and is of a slice type
// 	if list == nil || reflect.TypeOf(list).Kind() != reflect.Slice {
// 		return false
// 	}

// 	// Use reflection to iterate over the slice
// 	listValue := reflect.ValueOf(list)
// 	for i := 0; i < listValue.Len(); i++ {
// 		if reflect.DeepEqual(listValue.Index(i).Interface(), elem) {
// 			return true
// 		}
// 	}

// 	return false
// }

func (this *Exchange) IsArray(a interface{}) bool {
	// return reflect.TypeOf(a).Kind() == reflect.Slice
	switch a.(type) {
	case []interface{}:
		return true
	case []string:
		return true
	case []int:
		return true
	case []int64:
		return true
	case []float64:
		return true
	case []map[string]interface{}:
		return true
	}
	return false
}

func (this *Exchange) IndexBy(a interface{}, key interface{}) map[string]interface{} {
	outDict := make(map[string]interface{})
	var targetX []interface{}

	// Check if `a` is a slice of `[]interface{}` or a map
	if aArr, ok := a.([]interface{}); ok {
		targetX = aArr
	} else if aMap, ok := a.(map[string]interface{}); ok {
		for _, v := range aMap {
			targetX = append(targetX, v)
		}
	} else {
		return outDict // Unsupported type
	}

	// Process the slice `targetX`
	for _, elem := range targetX {
		switch v := elem.(type) {
		case map[string]interface{}:
			// Handle map entries
			if val, ok := v[ToString(key)]; ok {
				outDict[ToString(val)] = v
			}
		case []interface{}:
			// Handle slices of []interface{}
			if idx, ok := key.(int); ok && idx >= 0 && idx < len(v) {
				if keyStr, ok := v[idx].(string); ok {
					outDict[keyStr] = v
				}
			}
		case []string:
			// Handle slices of []string
			if idx, ok := key.(int); ok && idx >= 0 && idx < len(v) {
				outDict[v[idx]] = v
			}
		case []int:
			// Handle slices of []int
			if idx, ok := key.(int); ok && idx >= 0 && idx < len(v) {
				outDict[fmt.Sprintf("%d", v[idx])] = v
			}
		}
	}

	return outDict
}

// func (this *Exchange) IndexBy(a interface{}, key2 interface{}) map[string]interface{} {
// 	outDict := make(map[string]interface{})
// 	var targetX []interface{}
// 	if aArr, ok := a.([]interface{}); ok {
// 		targetX = aArr
// 	} else {
// 		for _, v := range a.(map[string]interface{}) {
// 			targetX = append(targetX, v)
// 		}
// 	}
// 	for _, elem := range targetX {
// 		if reflect.TypeOf(elem).Kind() == reflect.Map {
// 			elem2 := elem.(map[string]interface{})
// 			if val, ok := elem2[ToString(key2)]; ok {
// 				outDict[ToString(val)] = elem2
// 			}
// 		} else if reflect.TypeOf(elem).Kind() == reflect.Slice {
// 			index := key2.(int)
// 			elem2 := elem.([]interface{})
// 			if len(elem2) > index {
// 				outDict[elem2[index].(string)] = elem2
// 			}
// 		}
// 	}
// 	return outDict
// }

func (this *Exchange) GroupBy(trades interface{}, key2 interface{}) map[string]interface{} {
	key := key2.(string)
	outDict := make(map[string]interface{})
	list := trades.([]interface{})
	for _, elem := range list {
		elemDict := elem.(map[string]interface{})
		if val, ok := elemDict[key]; ok {
			elem2 := val.(string)
			if list2, exists := outDict[elem2]; exists {
				list2 = append(list2.([]interface{}), elem)
				outDict[elem2] = list2
			} else {
				outDict[elem2] = []interface{}{elem}
			}
		}
	}
	return outDict
}

func (this *Exchange) OmitZero(value interface{}) interface{} {
	switch v := value.(type) {
	case float64:
		if v == 0.0 {
			return nil
		}
	case int64:
		if v == 0 {
			return nil
		}
	case string:
		// Attempt to parse the string as a float64
		parsed, err := strconv.ParseFloat(v, 64)
		if err == nil {
			// If parsed value is zero, return nil
			if parsed == 0 {
				return nil
			}
		}
		// If parsing fails or value is non-zero, return the original string
	}
	return value
}

func (this *Exchange) Sum(args ...interface{}) interface{} {
	var res interface{} = 0.0
	for _, arg := range args {
		res = this.sumValues(res, arg)
	}
	return res
}

func (this *Exchange) sumValues(a, b interface{}) interface{} {
	return Add(a, b)
}

File: cs/ccxt/base/Exchange.Generic.cs
namespace ccxt;

using System.Globalization;

using dict = IDictionary<string, object>;
using list = List<object>;

public partial class BaseExchange
{
    public List<object> sortBy(object array, object value1, object desc2 = null, object defaultValue2 = null)
    {
        desc2 ??= false;
        var defaultValue = defaultValue2 ?? 0; // ts parity: generic.ts defaults the sort fallback to 0, not the empty string
        var desc = (bool)desc2;
        var list = (IList<object>)array;

        if (value1.GetType() == typeof(string))
        {
            var sortedList2 = list.OrderBy(x => ((dict)x)[(string)value1], sortKeyComparer).ToList();
            if (desc)
                sortedList2.Reverse();
            return sortedList2;
        }
        else
        {
            var value = (int)value1;
            var sortedList = list.OrderBy(x =>
            {
                if (x.GetType() == typeof(list))
                {
                    return ((list)x)[value];
                }
                return defaultValue;
            }, sortKeyComparer).ToList();

            if (desc)
                sortedList.Reverse();

            return sortedList;
        }
    }

    // a TOTAL comparator for sort keys, matching go's compareSortValues and
    // java's compareJsLike: numeric pairs compare as double regardless of how
    // they were boxed (Double from parsers next to Int32/Int64 from literals
    // and integer math used to throw ArgumentException out of
    // Double.CompareTo), and anything else falls back to ordinal string
    // comparison, so no key mix can throw. caveat shared with the double
    // route in go and java: integral keys above 2^53 can collide in double
    // space and settle on input order - ccxt's real keys (ms timestamps,
    // prices) sit well inside the exact range
    private class SortKeyComparer : IComparer<object>
    {
        private static bool isNumeric(object key)
        {
            return key is sbyte || key is byte || key is short || key is ushort || key is int || key is uint || key is long || key is ulong || key is float || key is double || key is decimal;
        }

        public int Compare(object a, object b)
        {
            if (isNumeric(a) && isNumeric(b))
            {
                return Convert.ToDouble(a).CompareTo(Convert.ToDouble(b));
            }
            var sa = (a == null) ? "" : a.ToString();
            var sb = (b == null) ? "" : b.ToString();
            return string.CompareOrdinal(sa, sb);
        }
    }

    private static readonly SortKeyComparer sortKeyComparer = new SortKeyComparer();

    public List<object> sortBy2(object array, object key1, object key2, object desc2 = null)
    {
        // todo :: check this later
        desc2 ??= false;
        var desc = (bool)desc2;
        var list = (IList<object>)array;

        if (key1.GetType() == typeof(string))
        {
            var orderByResult = list.OrderBy(s => ((dict)s)[(string)key1], sortKeyComparer)
                                    .ThenBy(s => ((dict)s)[(string)key2], sortKeyComparer)
                                    .ToList();
            if (desc)
                orderByResult.Reverse();
            return orderByResult;
        }
        return null;
    }

    public List<object> filterBy(object aa, object key, object value)
    {
        // var targetA = (List<object>)aa;
        var targetA = new List<object>() { };
        if (aa is List<object> plain)
        {
            targetA = plain;
        }
        else if (aa is dict asDict)
        {
            targetA = asDict.Values.ToList();
        }
        else if (aa is System.Collections.IEnumerable rows)
        {
            // a typed core's List<Dictionary<string, object>> (dydx FetchTransactionsHelper)
            foreach (var row in rows)
            {
                targetA.Add(row);
            }
        }
        var outList = new List<object>();
        foreach (object elem in targetA)
        {
            // JS reads a missing key as undefined and simply does not match; indexing the
            // dictionary directly threw KeyNotFoundException on entries lacking the key.
            var row = (dict)elem;
            object cell = row.TryGetValue((string)key, out var found) ? found : null;
            if (cell?.ToString() == value?.ToString())
            {
                outList.Add(elem);
            }
        }
        return outList;
    }

    public Dictionary<string, object> extend(object aa, object bb = null)
    {
        return Extend(aa, bb);
    }

    public static Dictionary<string, object> Extend(object aa, object bb = null)
    {

        var a = (dict)aa;
        var outDict = new Dictionary<string, object>();
        var keysA = new List<string>(a.Keys);
        foreach (string key in keysA)
            outDict[(string)key] = a[key];

        if (bb != null)
        {
            var b = (dict)bb;
            var keysB = new List<string>(b.Keys);
            foreach (string key in keysB)
                outDict[(string)key] = b[key];
        }
        return outDict;
    }

    public object deepExtend2(params object[] objs)
    {
        // old implementation
        object outDict = new Dictionary<string, object>();
        foreach (object obj in objs)
        {
            var obj2 = obj;
            if (obj2 == null)
            {
                obj2 = new Dictionary<string, object>();
            }
            if (obj2 is dict)
            {
                var keys = new List<string>(((dict)obj2).Keys);
                foreach (string key in keys)
                {

                    var value = ((dict)obj2)[key];
                    if (value != null && value is dict)
                    {
                        if (((dict)outDict).ContainsKey(key))
                        {
                            ((dict)outDict)[key] = deepExtend2(((dict)outDict)[key], value);
                        }
                        else
                        {
                            ((dict)outDict)[key] = deepExtend2(value);
                        }
                    }
                    else
                    {
                        ((dict)outDict)[key] = value;
                    }
                }
            }
            else
            {
                outDict = obj;
            }
        }
        return outDict;
    }
    public Dictionary<string, object> deepExtend(params object[] objs)
    {
        object outObj = null;
        foreach (object x in objs)
        {
            if (x == null)
                continue;

            if (x is dict)
            {
                if (outObj == null || !(outObj is dict))
                    outObj = new Dictionary<string, object>();

                var dictX = (dict)x;
                var dictXKeys = new List<string>(dictX.Keys);
                foreach (string k in dictXKeys)
                {
                    var arg1 = ((dict)outObj).ContainsKey(k) ? ((dict)outObj)[k] : null;
                    var arg2 = dictX.ContainsKey(k) ? dictX[k] : null;
                    // ((dict)outObj)[(string)k] = deepExtend(arg1, arg2); // this screws the  return type
                    if (arg1 is dict && arg2 is dict)
                    {
                        ((dict)outObj)[k] = deepExtend(arg1, arg2);
                    }
                    else
                    {
                        ((dict)outObj)[k] = arg2;
                    }
                }
            }
            else
            {
                outObj = x;
            }
        }
        return outObj as Dictionary<string, object>;
    }
    public bool inArray(object elem, object list2)
    {
        if (list2 == null)
            return false;
        if (list2 is List<string>)
        {
            return ((List<string>)list2).Contains((string)elem);
        }
        if (list2 is List<Int64>)
        {
            return ((List<Int64>)list2).Contains(Convert.ToInt64(elem));
        }
        var list = (List<object>)list2;
        if (elem is Int32 || elem is Int64)
        {
            return list.Contains(Convert.ToInt64(elem)) || list.Contains(Convert.ToInt32(elem));
        }
        return list.Contains(elem);
    }


    public bool isArray(object a)
    {
        return a is IList<object>;
    }

    public Dictionary<string, object> indexBySafe(object a, object key2)
    {
        return indexBy(a, key2); // this is needed for go
    }

    public Dictionary<string, object> indexBy(object a, object key2)
    {
        // var key = (string)key2;
        // var outDict = new Dictionary<string, object>();

        // List<object> input = null;
        // if (a.GetType() == typeof(List<object>))
        // {
        //     input = (List<object>)a;
        // }
        // else if (a.GetType() == typeof(List<string>))
        // {
        //     input = ((List<string>)a).Select(s => (object)s).ToList();
        // }
        // else if (a.GetType() == (typeof(dict)))
        // {
        //     input = new List<object>();
        //     foreach (string key3 in ((dict)a).Keys)
        //     {
        //         input.Add(((dict)a)[key3]);
        //     }
        // }

        // foreach (object elem in input)
        // {
        //     var elem2 = (dict)elem;
        //     if (elem2.ContainsKey(key))
        //     {
        //         // outDict.Add(((dict)elem)[key].ToString(), elem);
        //         outDict[(((dict)elem)[key]).ToString()] = elem2;
        //     }
        // }
        // return outDict;
        var outDict = new Dictionary<string, object>();
        var targetX = new List<object>() { };
        if (a.GetType() == typeof(List<object>))
        {
            targetX = (List<object>)a;
        }
        else
        {
            targetX = ((dict)a).Values.ToList();
        }
        foreach (object elem in targetX)
        {
            // var elem2 = (dict)elem;
            if (elem is dict)
            {
                var elem2 = (dict)elem;
                if (elem2.ContainsKey((string)key2))
                {
                    var elem2Value = elem2[(string)key2];
                    if (elem2Value == null)
                        continue;
                    outDict[elem2Value.ToString()] = elem2;
                }
            }
            else if (elem.GetType() == typeof(List<string>) || elem.GetType() == typeof(List<object>))
            {
                var index = Convert.ToInt32(key2);
                if (elem.GetType() == typeof(List<string>))
                {
                    var elem2 = (List<string>)elem;
                    if (elem2.Count > 0)
                    {
                        var elem2IndexValue = elem2[index];
                        if (elem2IndexValue == null)
                            continue;
                        outDict[elem2IndexValue.ToString()] = elem2;
                    }
                }
                if (elem.GetType() == typeof(List<object>))
                {
                    var elem2 = (List<object>)elem;
                    if (elem2.Count > 0)
                    {
                        var elem2IndexValue = elem2[index];
                        if (elem2IndexValue == null)
                            continue;
                        outDict[elem2IndexValue.ToString()] = elem2;
                    }
                }

            }

        }
        return outDict;
    }

    public Dictionary<string, object> groupBy(object trades, object key2)
    {
        var key = (string)key2;
        var outDict = new Dictionary<string, object>();
        var list = (List<object>)trades;
        foreach (object elem in list)
        {
            var elemDict = (dict)elem;
            if (elemDict.ContainsKey(key))
            {
                var elem2 = (string)elemDict[key];
                if (elem2 == null)
                    continue;
                if (outDict.ContainsKey(elem2))
                {
                    var list2 = (List<object>)outDict[elem2];
                    list2.Add(elem);
                    outDict[elem2] = list2;
                }
                else
                {
                    var list2 = new List<object>();
                    list2.Add(elem);
                    outDict[elem2] = list2;
                }

            }

        }

        return outDict;
    }

    public object omitZero(object value)
    {
        try
        {
            if (value is double)
            {
                if ((double)value == 0.0)
                {
                    return null;
                }
            }
            if (value is Int64)
            {
                if ((Int64)value == 0)
                {
                    return null;
                }
            }
            if (value is string)
            {
                // if ((string)value == "0")
                // {
                //     return null;
                // }
                var parsed = Convert.ToDouble(value, CultureInfo.InvariantCulture);
                if (parsed == 0)
                {
                    return null;
                }
            }
            return value;
        }
        catch (Exception e)
        {
            return value;
        }

    }

    public virtual object isDictionary(object value)
    {
        return isTrue(isTrue((!isEqual(value, null))) && isTrue(((value is IDictionary<string, object>)))) && !isTrue(((value is IList<object>) || (value.GetType().IsGenericType && value.GetType().GetGenericTypeDefinition().IsAssignableFrom(typeof(List<>)))));
    }

    public virtual object sum(params object[] args)
    {
        object res = 0;
        foreach (var arg in args)
        {
            res = sum(res, arg);
        }
        return res;
    }

    public virtual object sum(object a, object b)
    {
        if (a == null)
            a = 0;
        if (b == null)
            b = 0;
        var sum = Convert.ToDouble(a) + Convert.ToDouble(b);
        if (IsInteger(sum))
        {
            return Convert.ToInt64(sum);
        }
        return sum;
    }

}
